Core/BookmarksCoreDataStorage.swift

```diff
@@ -0,0 +1,722 @@
+    
+    public func containsBookmark(url: URL, completion: @escaping (Bool) -> Void) {
+        containsBookmark(url: url, searchType: .bookmarksOnly, completion: completion)
+    }
+    
+    public func containsFavorite(url: URL, completion: @escaping (Bool) -> Void) {
+        containsBookmark(url: url, searchType: .favoritesOnly, completion: completion)
+    }
+    
+    public func bookmark(forURL url: URL, completion: @escaping (Bookmark?) -> Void) {
+        viewContext.perform {
+    
+            let fetchRequest = NSFetchRequest<BookmarkManagedObject>(entityName: Constants.bookmarkClassName)
+            fetchRequest.predicate = NSPredicate(format: "url == %@ AND isFavorite == false", url as CVarArg)
+            
+            let results = try? self.viewContext.fetch(fetchRequest)
+            completion(results?.first)
+        }
+    }
+    
+    public func favorite(forURL url: URL, completion: @escaping (Bookmark?) -> Void) {
+        viewContext.perform {
+    
+            let fetchRequest = NSFetchRequest<BookmarkManagedObject>(entityName: Constants.bookmarkClassName)
+            fetchRequest.predicate = NSPredicate(format: "url == %@ AND isFavorite == true", url as CVarArg)
+            
+            let results = try? self.viewContext.fetch(fetchRequest)
+            completion(results?.first)
+        }
+    }
+    
+    // Just used for favicon deletion
+    public func allBookmarksAndFavoritesShallow(completion: @escaping ([Bookmark]) -> Void) {
+        viewContext.perform { [weak self] in
+        
+            let fetchRequest: NSFetchRequest<BookmarkManagedObject> = BookmarkManagedObject.fetchRequest()
+
+            guard let results = try? self?.viewContext.fetch(fetchRequest) else {
+                assertionFailure("Error fetching Bookmarks")
+                return
+            }
+            
+            completion(results)
+        }
+    }
+    
+    public func saveNewFolder(withTitle title: String, parentID: NSManagedObjectID, completion: BookmarkItemSavedCompletion? = nil) {
+        createFolder(title: title, isFavorite: false, parentID: parentID, completion: completion)
+    }
+    
+    public func saveNewFavorite(withTitle title: String, url: URL, completion: BookmarkItemSavedCompletion? = nil) {
+        createBookmark(url: url, title: title, isFavorite: true, completion: completion)
+    }
+    
+    public func saveNewBookmark(withTitle title: String, url: URL, parentID: NSManagedObjectID?, completion: BookmarkItemSavedCompletion? = nil) {
+        createBookmark(url: url, title: title, isFavorite: false, parentID: parentID, completion: completion)
+    }
+    
+    // Since we'll never need to update children at the same time, we only support changing the title and parent
+    public func update(folderID: NSManagedObjectID,
+                       newTitle: String,
+                       newParentID: NSManagedObjectID,
+                       completion: BookmarkItemUpdatedCompletion? = nil) {
```

Comment:
I don't think any objects that are retrieved from this API are tied to any particular thread - or am missing something?

Reply:
I mean with e.g. `public func bookmark(forURL url: URL, completion: @escaping (Bookmark?) -> Void)`, it's completion will be invoked on the main thread, which sort of makes sense to me since the object is from the view context. Doesn't the fact that it's from the view context (and e.g. if I tried to check the parent or child and it's faulted on a background thread it could crash) mean we have to make assumptions about the API?

Reply:
Sorry, I got totally confused looking at the API. My mind got the impression that `Bookmark` is not a managed object - and thus it is not tied to any thread.

Now I have few trains of thought:
1. We could make API more explicit about what is returned: seeing `BookmarkManagedObject` in the completion would make people think about thread affiliation of particular instances. It would also disclose the information about inner implementation, which in this case is not bad as API has limitation around threading related to Core Data. I know it's not ideal as `Bookmark` abstraction is elegant, but at the same time it does not warn the user of this API about its limitations (I've assumed it's a thread-safe struct 🤦‍♂️and was ready to just pass it between threads... )
2. For methods that return object IDs in completion, I'd consider if we want to make it also return on the Main Thread, to keep API consistent. That would tie the whole object to Main Thread, but not in the performance-impairing way.
3. If users of this API understand relation to Main thread (based on above two points), then it is less likely any locking they implement won't take that into account - DispatchGroup from the above would have to be transformed to be async call based on that.

Does that makes sense to you?

Reply:
Yeah, that does make sense. Agreed on #1, I'd rather we had the abstraction, but, well, we just don't as long as we're using core data properly.
Do you think I should change BookmarksManager to return a `BookmarkManagedObject` instead too? 
2 also makes sense, but it still wouldn't feel super obvious from the outside to me. How would you feel about something in the completion type alias name to indicate the thread?

Reply:
Yes to all of your points, we should make it as clear as possible. :)